// @react-vertex/core/dist/cjs/sceneGraph/InstancedNode.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InstancedNode = exports.instancedExt = exports.isInstancedNode = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _GraphNode = _interopRequireDefault(require("./GraphNode"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var isInstancedNode = Symbol('isInstancedNode');
exports.isInstancedNode = isInstancedNode;
var instancedExt = 'ANGLE_instanced_arrays';
exports.instancedExt = instancedExt;

var InstancedNode = /*#__PURE__*/function (_Node) {
  (0, _inherits2["default"])(InstancedNode, _Node);

  var _super = _createSuper(InstancedNode);

  function InstancedNode() {
    var _this;

    (0, _classCallCheck2["default"])(this, InstancedNode);
    _this = _super.call(this);
    _this.index = void 0;
    _this.attributes = void 0;
    _this.drawElements = void 0;
    _this[isInstancedNode] = true;
    _this.index = 0;
    _this.attributes = {};
    return _this;
  }

  (0, _createClass2["default"])(InstancedNode, [{
    key: "applyProps",
    value: function applyProps(nextprops) {
      var prevProps = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      this.applyMatrixProps(nextprops, prevProps);

      if (nextprops.index !== undefined) {
        this.index = nextprops.index;
      }

      if (nextprops.attributes !== undefined) {
        this.attributes = nextprops.attributes;
      }

      this.drawElements = nextprops.drawElements;
    }
  }]);
  return InstancedNode;
}(_GraphNode["default"]);

exports.InstancedNode = InstancedNode;
