// @react-vertex/core/dist/esm/hooks/uniforms.js
import { useMemo, useEffect } from 'react';
import { useSceneNode } from '..';
export function useUniformSampler2d(gl, program, name, texture) {
  var scene = useSceneNode();
  var unit = useMemo(function () {
    return scene.getTextureUnit(texture);
  }, [scene, texture]);
  var memoized = useMemo(function () {
    var location = gl.getUniformLocation(program, name);
    gl.uniform1i(location, unit);
    return location;
  }, [gl, program, name, unit]);
  useEffect(function () {
    return function () {
      return scene.releaseTextureUnit(unit);
    };
  }, [scene, unit]);
  return [unit, memoized];
} // *************************************************
// Matrix Uniforms
// *************************************************

function useMatrixUniform(gl, program, name, value, type) {
  var memoized = useMemo(function () {
    return gl.getUniformLocation(program, name);
  }, [gl, program, name]);
  useEffect(function () {
    gl.useProgram(program);
    gl[type](memoized, false, value);
  }, [gl, program, type, memoized, value]);
  return memoized;
}

export function useUniformMatrix2fv(gl, program, name, value) {
  return useMatrixUniform(gl, program, name, value, 'uniformMatrix2fv');
}
export function useUniformMatrix3fv(gl, program, name, value) {
  return useMatrixUniform(gl, program, name, value, 'uniformMatrix3fv');
}
export function useUniformMatrix4fv(gl, program, name, value) {
  return useMatrixUniform(gl, program, name, value, 'uniformMatrix4fv');
} // *************************************************
// Number Uniforms
// *************************************************

function useNumberUniform(gl, program, name, value, type) {
  var memoized = useMemo(function () {
    return gl.getUniformLocation(program, name);
  }, [gl, program, name]);
  useEffect(function () {
    gl.useProgram(program);
    gl[type](memoized, value);
  }, [gl, program, type, memoized, value]);
  return memoized;
}

export function useUniform1f(gl, program, name, value) {
  return useNumberUniform(gl, program, name, value, 'uniform1f');
}
export function useUniform1i(gl, program, name, value) {
  return useNumberUniform(gl, program, name, value, 'uniform1i');
} // *************************************************
// Vector Uniforms
// *************************************************

function useVectorUniform(gl, program, name, value, type) {
  var memoized = useMemo(function () {
    return gl.getUniformLocation(program, name);
  }, [gl, program, name]);
  useEffect(function () {
    gl.useProgram(program);
    gl[type](memoized, value);
  }, [gl, program, type, memoized, value]);
  return memoized;
}

export function useUniform1fv(gl, program, name, value) {
  return useVectorUniform(gl, program, name, value, 'uniform1fv');
}
export function useUniform1iv(gl, program, name, value) {
  return useVectorUniform(gl, program, name, value, 'uniform1iv');
}
export function useUniform2fv(gl, program, name, value) {
  return useVectorUniform(gl, program, name, value, 'uniform2fv');
}
export function useUniform2iv(gl, program, name, value) {
  return useVectorUniform(gl, program, name, value, 'uniform2iv');
}
export function useUniform3fv(gl, program, name, value) {
  return useVectorUniform(gl, program, name, value, 'uniform3fv');
}
export function useUniform3iv(gl, program, name, value) {
  return useVectorUniform(gl, program, name, value, 'uniform3iv');
}
export function useUniform4fv(gl, program, name, value) {
  return useVectorUniform(gl, program, name, value, 'uniform4fv');
}
export function useUniform4iv(gl, program, name, value) {
  return useVectorUniform(gl, program, name, value, 'uniform4iv');
}
