// @react-vertex/core/dist/cjs/hooks/framebuffers.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useFramebuffer = useFramebuffer;
exports.useFBO = useFBO;
exports.useDoubleFBO = useDoubleFBO;

var _react = require("react");

var _textures = require("./textures");

var _warning = _interopRequireDefault(require("warning"));

var prefix = 'react-vertex:';

function useFramebuffer(gl) {
  var memoized = (0, _react.useMemo)(function () {
    var buffer = gl.createFramebuffer();
    (0, _warning["default"])(!!buffer, "".concat(prefix, " Failed to create frame buffer."));
    return buffer;
  }, [gl]);
  (0, _react.useEffect)(function () {
    return function () {
      return gl.deleteFramebuffer(memoized);
    };
  }, [gl, memoized]);
  return memoized;
}

function useFBO(gl, width, height, getTexOpts) {
  var tex = (0, _textures.useDataTexture)(gl, null, width, height, getTexOpts);
  var fbo = useFramebuffer(gl);
  var memoized = (0, _react.useMemo)(function () {
    var attachment = gl.COLOR_ATTACHMENT0;
    gl.bindFramebuffer(gl.FRAMEBUFFER, fbo);
    gl.framebufferTexture2D(gl.FRAMEBUFFER, attachment, gl.TEXTURE_2D, tex, 0);
    return {
      tex: tex,
      fbo: fbo,
      attach: function attach(unit) {
        gl.activeTexture(gl.TEXTURE0 + unit);
        gl.bindTexture(gl.TEXTURE_2D, tex);
        return unit;
      }
    };
  }, [gl, fbo, tex]);
  return memoized;
}

function useDoubleFBO(gl, width, height, getTexOpts) {
  var frameBuffer1 = useFBO(gl, width, height, getTexOpts);
  var frameBuffer2 = useFBO(gl, width, height, getTexOpts);
  var memoized = (0, _react.useMemo)(function () {
    var fbo1 = frameBuffer1;
    var fbo2 = frameBuffer2;
    return {
      get read() {
        return fbo1;
      },

      set read(value) {
        fbo1 = value;
      },

      get write() {
        return fbo2;
      },

      set write(value) {
        fbo2 = value;
      },

      swap: function swap() {
        var temp = fbo1;
        fbo1 = fbo2;
        fbo2 = temp;
      }
    };
  }, [frameBuffer1, frameBuffer2]);
  return memoized;
}
