// @react-vertex/core/dist/cjs/hooks/uniforms.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useUniformSampler2d = useUniformSampler2d;
exports.useUniformMatrix2fv = useUniformMatrix2fv;
exports.useUniformMatrix3fv = useUniformMatrix3fv;
exports.useUniformMatrix4fv = useUniformMatrix4fv;
exports.useUniform1f = useUniform1f;
exports.useUniform1i = useUniform1i;
exports.useUniform1fv = useUniform1fv;
exports.useUniform1iv = useUniform1iv;
exports.useUniform2fv = useUniform2fv;
exports.useUniform2iv = useUniform2iv;
exports.useUniform3fv = useUniform3fv;
exports.useUniform3iv = useUniform3iv;
exports.useUniform4fv = useUniform4fv;
exports.useUniform4iv = useUniform4iv;

var _react = require("react");

var _ = require("..");

function useUniformSampler2d(gl, program, name, texture) {
  var scene = (0, _.useSceneNode)();
  var unit = (0, _react.useMemo)(function () {
    return scene.getTextureUnit(texture);
  }, [scene, texture]);
  var memoized = (0, _react.useMemo)(function () {
    var location = gl.getUniformLocation(program, name);
    gl.uniform1i(location, unit);
    return location;
  }, [gl, program, name, unit]);
  (0, _react.useEffect)(function () {
    return function () {
      return scene.releaseTextureUnit(unit);
    };
  }, [scene, unit]);
  return [unit, memoized];
} // *************************************************
// Matrix Uniforms
// *************************************************


function useMatrixUniform(gl, program, name, value, type) {
  var memoized = (0, _react.useMemo)(function () {
    return gl.getUniformLocation(program, name);
  }, [gl, program, name]);
  (0, _react.useEffect)(function () {
    gl.useProgram(program);
    gl[type](memoized, false, value);
  }, [gl, program, type, memoized, value]);
  return memoized;
}

function useUniformMatrix2fv(gl, program, name, value) {
  return useMatrixUniform(gl, program, name, value, 'uniformMatrix2fv');
}

function useUniformMatrix3fv(gl, program, name, value) {
  return useMatrixUniform(gl, program, name, value, 'uniformMatrix3fv');
}

function useUniformMatrix4fv(gl, program, name, value) {
  return useMatrixUniform(gl, program, name, value, 'uniformMatrix4fv');
} // *************************************************
// Number Uniforms
// *************************************************


function useNumberUniform(gl, program, name, value, type) {
  var memoized = (0, _react.useMemo)(function () {
    return gl.getUniformLocation(program, name);
  }, [gl, program, name]);
  (0, _react.useEffect)(function () {
    gl.useProgram(program);
    gl[type](memoized, value);
  }, [gl, program, type, memoized, value]);
  return memoized;
}

function useUniform1f(gl, program, name, value) {
  return useNumberUniform(gl, program, name, value, 'uniform1f');
}

function useUniform1i(gl, program, name, value) {
  return useNumberUniform(gl, program, name, value, 'uniform1i');
} // *************************************************
// Vector Uniforms
// *************************************************


function useVectorUniform(gl, program, name, value, type) {
  var memoized = (0, _react.useMemo)(function () {
    return gl.getUniformLocation(program, name);
  }, [gl, program, name]);
  (0, _react.useEffect)(function () {
    gl.useProgram(program);
    gl[type](memoized, value);
  }, [gl, program, type, memoized, value]);
  return memoized;
}

function useUniform1fv(gl, program, name, value) {
  return useVectorUniform(gl, program, name, value, 'uniform1fv');
}

function useUniform1iv(gl, program, name, value) {
  return useVectorUniform(gl, program, name, value, 'uniform1iv');
}

function useUniform2fv(gl, program, name, value) {
  return useVectorUniform(gl, program, name, value, 'uniform2fv');
}

function useUniform2iv(gl, program, name, value) {
  return useVectorUniform(gl, program, name, value, 'uniform2iv');
}

function useUniform3fv(gl, program, name, value) {
  return useVectorUniform(gl, program, name, value, 'uniform3fv');
}

function useUniform3iv(gl, program, name, value) {
  return useVectorUniform(gl, program, name, value, 'uniform3iv');
}

function useUniform4fv(gl, program, name, value) {
  return useVectorUniform(gl, program, name, value, 'uniform4fv');
}

function useUniform4iv(gl, program, name, value) {
  return useVectorUniform(gl, program, name, value, 'uniform4iv');
}
