// @react-vertex/core/dist/esm/sceneGraph/InstancedNode.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import Node from './GraphNode';
export var isInstancedNode = Symbol('isInstancedNode');
export var instancedExt = 'ANGLE_instanced_arrays';
export var InstancedNode = /*#__PURE__*/function (_Node) {
  _inherits(InstancedNode, _Node);

  var _super = _createSuper(InstancedNode);

  function InstancedNode() {
    var _this;

    _classCallCheck(this, InstancedNode);

    _this = _super.call(this);
    _this.index = void 0;
    _this.attributes = void 0;
    _this.drawElements = void 0;
    _this[isInstancedNode] = true;
    _this.index = 0;
    _this.attributes = {};
    return _this;
  }

  _createClass(InstancedNode, [{
    key: "applyProps",
    value: function applyProps(nextprops) {
      var prevProps = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      this.applyMatrixProps(nextprops, prevProps);

      if (nextprops.index !== undefined) {
        this.index = nextprops.index;
      }

      if (nextprops.attributes !== undefined) {
        this.attributes = nextprops.attributes;
      }

      this.drawElements = nextprops.drawElements;
    }
  }]);

  return InstancedNode;
}(Node);
