// @react-vertex/core/dist/cjs/Reconciler/hostConfig.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.hostConfig = void 0;

var _scheduler = require("scheduler");

var _warning = _interopRequireDefault(require("warning"));

var _sceneGraph = require("../sceneGraph");

// @ts-nocheck
var NO_CONTEXT = true;

var NOOP = function NOOP() {};

var addChild = function addChild(parentInstance, child) {
  if (typeof child === 'string') {
    (0, _warning["default"])(false, 'React Vertex does not support text nodes.');
    return;
  }

  parentInstance.add(child);
};

var hostConfig = {
  schedulePassiveEffects: _scheduler.unstable_scheduleCallback,
  cancelPassiveEffects: _scheduler.unstable_cancelCallback,
  now: Date.now,
  prepareForCommit: NOOP,
  resetAfterCommit: NOOP,
  getRootHostContext: function getRootHostContext() {
    return NO_CONTEXT;
  },
  getChildHostContext: function getChildHostContext() {
    return NO_CONTEXT;
  },
  shouldSetTextContent: function shouldSetTextContent(type, props) {
    return typeof props.children === 'string' || typeof props.children === 'number';
  },
  createInstance: function createInstance(type, props, internalInstanceHandle) {
    var instance;

    switch (type) {
      case 'group':
        instance = new _sceneGraph.GroupNode();
        instance.root = internalInstanceHandle;
        break;

      case 'camera':
        instance = new _sceneGraph.CameraNode();
        instance.root = internalInstanceHandle;
        break;

      case 'material':
        instance = new _sceneGraph.MaterialNode();
        instance.root = internalInstanceHandle;
        break;

      case 'geometry':
        instance = new _sceneGraph.GeometryNode();
        instance.root = internalInstanceHandle;
        break;

      case 'instancedgeometry':
        instance = new _sceneGraph.InstancedNode();
        instance.root = internalInstanceHandle;
        break;
    }

    (0, _warning["default"])(instance, 'React Vertex does not support the type "%s"', type);
    instance.applyProps(props);
    return instance;
  },
  createTextInstance: function createTextInstance(text) {
    return text;
  },
  resetTextContent: NOOP,
  getPublicInstance: function getPublicInstance(instance) {
    return instance;
  },
  appendChildToContainer: addChild,
  removeChildFromContainer: function removeChildFromContainer(parentInstance, child) {
    parentInstance.remove(child);
  },
  appendInitialChild: addChild,
  appendChild: addChild,
  removeChild: function removeChild(parentInstance, child) {
    parentInstance.remove(child);
  },
  insertBefore: addChild,
  insertInContainerBefore: addChild,
  finalizeInitialChildren: NOOP,
  supportsMutation: true,
  prepareUpdate: function prepareUpdate() {
    return true;
  },
  commitUpdate: function commitUpdate(instance, payload, type, oldProps, newProps) {
    instance.applyProps(newProps, oldProps);
  }
};
exports.hostConfig = hostConfig;
