// @react-vertex/core/dist/esm/Canvas.js
import _typeof from "@babel/runtime/helpers/typeof";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import React, { Component, createRef } from 'react';
import ReactVertexReconciler, { SceneNode } from './Reconciler';
import ReactVertexContext from './Context';

var Canvas = /*#__PURE__*/function (_Component) {
  _inherits(Canvas, _Component);

  var _super = _createSuper(Canvas);

  function Canvas() {
    var _this;

    _classCallCheck(this, Canvas);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this.sceneNode = void 0;
    _this.state = {
      error: false,
      message: ''
    };
    _this.canvas = /*#__PURE__*/createRef();
    _this.container = void 0;
    _this.contextObject = void 0;

    _this.renderScene = function () {
      if (_this.sceneNode) {
        _this.sceneNode.render();
      }
    };

    return _this;
  }

  _createClass(Canvas, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var current = this.canvas.current;
      var _this$props = this.props,
          _this$props$webgl = _this$props.webgl1,
          webgl1 = _this$props$webgl === void 0 ? true : _this$props$webgl,
          _this$props$webgl2 = _this$props.webgl2,
          webgl2 = _this$props$webgl2 === void 0 ? false : _this$props$webgl2,
          _this$props$clearColo = _this$props.clearColor,
          clearColor = _this$props$clearColo === void 0 ? [0, 0, 0, 1] : _this$props$clearColo,
          children = _this$props.children,
          _this$props$antialias = _this$props.antialias,
          antialias = _this$props$antialias === void 0 ? false : _this$props$antialias,
          _this$props$textureFl = _this$props.textureFlip,
          textureFlip = _this$props$textureFl === void 0 ? true : _this$props$textureFl,
          _this$props$contextAt = _this$props.contextAttrs,
          contextAttrs = _this$props$contextAt === void 0 ? {} : _this$props$contextAt,
          _this$props$extension = _this$props.extensions,
          extensions = _this$props$extension === void 0 ? [] : _this$props$extension,
          _this$props$renderOnU = _this$props.renderOnUpdate,
          renderOnUpdate = _this$props$renderOnU === void 0 ? false : _this$props$renderOnU;

      var attrs = _objectSpread({
        antialias: antialias
      }, contextAttrs);

      if (!current) {
        return;
      }

      var gl, webglVersion;

      if (webgl2) {
        gl = current.getContext('webgl2', attrs);
        webglVersion = 2;
      }

      if (!gl && webgl1) {
        gl = current.getContext('webgl', attrs);
        webglVersion = 1;
      }

      if (!gl) {
        this.setState({
          error: true,
          message: 'Could not create WebGL context.'
        });
        return;
      }

      textureFlip && gl.pixelStorei(gl.UNPACK_FLIP_Y_WEBGL, true);
      this.sceneNode = new SceneNode(current, extensions, gl);

      if (typeof webglVersion === 'number') {
        this.sceneNode.webglVersion = webglVersion;
      }

      this.sceneNode.clearColor = clearColor;
      this.sceneNode.renderOnUpdate = renderOnUpdate;

      if ((typeof window === "undefined" ? "undefined" : _typeof(window)) === 'object') {
        // @ts-ignore
        window.sceneNode = this.sceneNode;
      }

      var update = this.updateDimensions();

      if (update) {
        this.contextObject = {
          scene: this.sceneNode,
          width: update.width,
          height: update.height
        };
      }

      this.container = ReactVertexReconciler.createContainer(this.sceneNode, false, false);

      if (this.contextObject) {
        ReactVertexReconciler.updateContainer( /*#__PURE__*/React.createElement(ReactVertexContext.Provider, {
          value: this.contextObject
        }, children), this.container, this, function () {});
      }
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      var children = this.props.children;
      var dims = this.updateDimensions();

      if (dims && dims.update && this.contextObject) {
        this.contextObject = _objectSpread(_objectSpread({}, this.contextObject), {}, {
          width: dims.width,
          height: dims.height
        });
      }

      if (this.container && this.contextObject) {
        ReactVertexReconciler.updateContainer( /*#__PURE__*/React.createElement(ReactVertexContext.Provider, {
          value: this.contextObject
        }, children), this.container, this, function () {});
      }
    }
  }, {
    key: "updateDimensions",
    value: function updateDimensions() {
      var current = this.canvas.current;

      if (!current) {
        return;
      }

      var _this$props2 = this.props,
          width = _this$props2.width,
          height = _this$props2.height,
          renderOnResize = _this$props2.renderOnResize;
      var devicePixelRatio = window.devicePixelRatio || 1;
      var nextWidth = Math.round(width * devicePixelRatio);
      var nextHeight = Math.round(height * devicePixelRatio);
      var update = nextWidth !== current.width || nextHeight !== current.height;

      if (update) {
        current.style.width = "".concat(width, "px");
        current.style.height = "".concat(height, "px");
        current.width = nextWidth;
        current.height = nextHeight;
        renderOnResize && this.sceneNode && this.sceneNode.requestRender();
      }

      return {
        width: nextWidth,
        height: nextHeight,
        update: update
      };
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (this.container) {
        ReactVertexReconciler.updateContainer(null, this.container, this, function () {});
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props3 = this.props,
          _this$props3$canvasCl = _this$props3.canvasClass,
          canvasClass = _this$props3$canvasCl === void 0 ? '' : _this$props3$canvasCl,
          _this$props3$canvasSt = _this$props3.canvasStyle,
          canvasStyle = _this$props3$canvasSt === void 0 ? {} : _this$props3$canvasSt;
      return /*#__PURE__*/React.createElement("canvas", {
        ref: this.canvas,
        className: canvasClass,
        style: canvasStyle
      });
    }
  }]);

  return Canvas;
}(Component);

export { Canvas as default };
